vcd/catalogitem.go

```diff
@@ -69,7 +78,49 @@ func findCatalogItem(d *schema.ResourceData, vcdClient *VCDClient, origin string
 		identifier = d.Get("name").(string)
 	}
 
-	catalogItem, err := catalog.GetCatalogItemByNameOrId(identifier, false)
+	var catalogItem *govcd.CatalogItem
+	if origin == "datasource" {
+		filter, hasFilter := d.GetOk("filter")
+		if hasFilter {
```

Comment:
hmm, I got impression what with govcd changes the code here will be much cleaner - if it possible this all code would be wrapped and be in govcd?
Here we just:
* buildCriteria from filter
* catalogItem = catalog.GetCatalogByFilter

Reply:
Thanks. This is a good suggestion. It should be doable. Going to have a look into it.

I wanted to have an uniformed way of running the searches for different resources, but in fact, the lack of generics in Go makes a true abstract approach unfeasible.

In summary, what we want is to have (in govcd)
* org.GetCatalogByFilter
* vdc.GetEdgeGatewayByFilter
* vdc.GetNetworkByFilter
* catalog.GetvAppTemplateByFilter
* catalog.GetCatalogMediaByFilter

@Didainius , @lvirbalas what do you think?

Reply:
Looks fine as long as we don't depend on internal Terraform structures in govcd.

Reply:
I made a new file (`filter_get.go`) in this same repository, and centralized all retrieval by filter functions there. This allows us to use just one call within the data source code, and delegate the unpacking of the filter structure to these functions, which in turn use a low-level function to convert the filter structure and do the search. All in all, the code is more readable, and probably better than moving all this part to govcd. 

I feel that govcd should offer the basics, and leave the specific usage (such as failing for multiple items in the findings) to the client, i.e. the terraform-provider-vcd plugin.